pkg/controller/controller.go

```diff
@@ -705,6 +684,53 @@ func (c *Controller) renderActionsetArtifacts(ctx context.Context,
 	}
 }
 
+func (c *Controller) maybeSetActionSetStateComplete(ctx context.Context,
+	as *crv1alpha1.ActionSet,
+	aIDX int,
+	bp *crv1alpha1.Blueprint,
+	coreErr, deferErr error,
+) {
+	af := func(ras *crv1alpha1.ActionSet) error {
+		// Running phase is in the current action, TODO: #2270 track multiple phases
+		if isPhaseInAction(ras.Status.Progress.RunningPhase, ras.Status.Actions[aIDX]) {
+			// set the RunningPhase to empty string
+			ras.Status.Progress.RunningPhase = ""
+		}
+
+		// make sure that the core phases that were run also didnt return any error
+		// and then set actionset's state to be complete
+		if coreErr != nil || deferErr != nil {
+			ras.Status.State = crv1alpha1.StateFailed
+			return nil
+		}
+
+		for _, as := range ras.Status.Actions {
+			for _, p := range as.Phases {
+				if p.State != crv1alpha1.StateComplete {
+					log.WithContext(ctx).Print(
+						"Finished action, but other action's phase is still running. Not setting state to complete.",
```

Comment:
I have a question about this, in the PR desc also it's mentioned that `Actionset state should be updated only when all actions complete by onUpdateActionSet`. 

What I am not able to understand is, the relation between two actions. 
`Finished action, but other action's phase is still running.` if backup action is complete should it not be marked complete? Which other action would be wait for before marking backup action to be complete?

Reply:
I was hoping we can make it work on `onUpdateActionSet`, but there are unresolved races there. I will change the description.

Reply:
> if backup action is complete should it not be marked complete

There is no status per action, adding which would be a bigger change than a bugfix. We have status per phase and per actionset currently. So we can't mark an action to be complete (which would make things a bit easier TBH).